### DISCUSSION TITLE
Precompiled ifcopenshell core library for linux?

### DISCUSSION THREAD

**Discussion by bdrnX** [IfcOpenShell/IfcOpenShell#5778] [Q&A]
**My goal:** Using the examples in 'src/examples', specially the alignment examples. I don't mind using python but there's no examples specially for alignment. 

**Problem:** failed to build the C++ library on linux :(. The error is at the end of the post
**Question 1:** building on my PC takes ages, is there another way to runing these examples? maybe using a pre-combiled...?
**Question 2:** are there Alignment examples in python?

Note:
According to the documentation, if I to use C++, I need to recompile the following repo:

`git clone --recursive https://github.com/IfcOpenShell/IfcOpenshell.git`

Shouldn't it be a tagged version! or this repo: `github.com/IfcOpenBot/IfcOpenShell`

**Error**:
```
/usr/bin/ld: CMakeFiles/IfcAdvancedHouse.dir/IfcAdvancedHouse.cpp.o: in function `main':
IfcAdvancedHouse.cpp:(.text.startup+0x552): undefined reference to `IfcGeom::serialise(std::__cxx11::basic_strin
g<char, std::char_traits<char>, std::allocator<char> > const&, TopoDS_Shape const&, bool)'
/usr/bin/ld: IfcAdvancedHouse.cpp:(.text.startup+0x687): undefined reference to `IfcGeom::serialise(std::__cxx11
::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, TopoDS_Shape const&, bool)'
/usr/bin/ld: IfcAdvancedHouse.cpp:(.text.startup+0x8ed): undefined reference to `IfcGeom::tesselate(std::__cxx11
::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, TopoDS_Shape const&, double)'
/usr/bin/ld: CMakeFiles/IfcOpenHouse.dir/IfcOpenHouse.cpp.o: in function `main':
IfcOpenHouse.cpp:(.text.startup+0x222e): undefined reference to `IfcGeom::tesselate(std::__cxx11::basic_string<c
har, std::char_traits<char>, std::allocator<char> > const&, TopoDS_Shape const&, double)'
collect2: error: ld returned 1 exit status
make[2]: *** [examples/CMakeFiles/IfcAdvancedHouse.dir/build.make:175: examples/IfcAdvancedHouse] Error 1
make[1]: *** [CMakeFiles/Makefile2:644: examples/CMakeFiles/IfcAdvancedHouse.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
collect2: error: ld returned 1 exit status
make[2]: *** [examples/CMakeFiles/IfcOpenHouse.dir/build.make:175: examples/IfcOpenHouse] Error 1
make[1]: *** [CMakeFiles/Makefile2:606: examples/CMakeFiles/IfcOpenHouse.dir/all] Error 2
make: *** [Makefile:156: all] Error 2


```

**Comment by aothms** [comment]
>  Question 1: building on my PC takes ages, is there another way to runing these examples? maybe using a pre-combiled...?

We don't plan to offer precompiled binaries at the moment.

If you're particularly interested in alignment, I would advise to compile only with the IFC4X3_ADD2 schema version. When running cmake: `-DSCHEMA_VERSIONS=4x3_add2`.

>  Question 2: are there Alignment examples in python?

The highest amount of complexity comes from working with the IFC schema itself. If you can figure out how to do that in C++, doing that in Python is only easier.

The community here can definitely help you get started, especially when you're willing to help writing those examples. The 4.3 schema is still rather new. Contributions very welcome :)

> Shouldn't it be a tagged version! or this repo: github.com/IfcOpenBot/IfcOpenShell

The other repo is only build automation. I'd advise to always use the latest commit, especially C++ does not change that often.

> IfcAdvancedHouse.cpp:(.text.startup+0x552): undefined reference to `IfcGeom::serialise(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, TopoDS_Shape const&, bool)'

You most likely miss the serializer libs (I just saw they're not being installed, fixed after 2f7412d0c, maybe that was the reason). `geometry_serializer geometry_serializer_ifc2x3 geometry_serializer_ifc4 geometry_serializer_ifc4x3_add2` (or whatever schemas you end up compiling into the lib).

**Comment by RickBrice** [comment]
Here is a link to information about alignment modeling that supports the C++ alignment examples. There is also a link in the write-up for a python example.

https://wiki.osarch.org/index.php?title=IFC_-_Industry_Foundation_Classes/IFC_alignment
